lib/features/build_server/cache/project_cache.dart

```diff
@@ -0,0 +1,235 @@
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:flyde/core/fs/configs/compiler_config.dart';
+import 'package:flyde/core/fs/file_extension.dart';
+import 'package:flyde/features/build_server/cache/implementation_object_ref.dart';
+import 'package:flyde/features/build_server/cache/lock/config_lock.dart';
+import 'package:flyde/features/build_server/cache/lock/project_cache_lock.dart';
+import 'package:flyde/core/fs/wrapper/source_file.dart';
+import 'package:flyde/features/build_server/cache/lock/source_file_lock.dart';
+import 'package:path/path.dart';
+
+/// The cache for a single project.
+///
+/// Creates an abstraction around the raw source and object files.
+///
+/// After an instance has been created `init` has to be called.
+/// ```
+/// final cache = ProjectCache('my-id', directoryWhereCacheIsStored);
+/// await cache.init(); // Neccessary to load persisted state
+/// ```
+class ProjectCache {
+  final String _projectId;
+
+  late final Directory _workingDirectory;
+
+  late final ProjectCacheLock _lock;
+
+  CompilerConfig? _config;
+
+  ProjectCache(this._projectId, Directory cacheDir) {
+    _workingDirectory = Directory(join(cacheDir.path, '$_projectId'));
+  }
+
+  /// Initiates the cache and loads persisted state if available.
+  Future<void> init() async {
+    if (await _lockFile.exists()) {
+      final content = await _lockFile.readAsString();
+
+      if (content.isEmpty) {
+        _lock = ProjectCacheLock(configs: {}, files: {});
+      } else {
+        _lock = ProjectCacheLock.fromJson(jsonDecode(content));
+      }
+    } else {
+      await _lockFile.create(recursive: true);
+      _lock = ProjectCacheLock(configs: {}, files: {});
+    }
+  }
+
+  Future<void> finish() async {
+    await _lockFile.writeAsString(json.encode(_lock.toJson()));
+  }
+
+  /// Synchronizes the cache with the user project.
+  ///
+  /// [files] is a map where each unique file id is associated to it's hash value.
+  /// `sync` checks which stored files are outdated and returns a `List` of
```

Comment:
'returns a list of those outdated files' -> You actually return the `required` files.

Reply:
outdated -> files which are stored in cache but have not the same content as the file in the user project